src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +113,252 @@ object RuleUtils {
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+    if (useBucketUnion) {
+      // if BucketSpec of the index is used to read the index data, we need to shuffle
+      // the appended data in the same way to avoid additional shuffle of index data.
+      getComplementIndexPlan(spark, index, plan, replacedPlan)
+    } else {
+      replacedPlan
+    }
+  }
+
+  /**
+   * Get complement plan for an index with appended data.
+   *
+   * This method consists of the following steps
+   * 1) Get a plan from originalPlan by replacing data location with appended data
+   * 2) On-the-fly shuffle for the appended data, using indexedColumns & numBuckets.
+   *   - Shuffle is located before Project to utilize Push-down Filters
+   *     - Shuffle => Project => Filter => Relation
+   *   - if Project filters indexedColumns, then Shuffle should be located after the node
+   *     - Project => Shuffle => Filter => Relation
+   * 3) Bucket-aware union both indexPlan and complementPlan to avoid repartitioning index data
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session
+   * @param index index used in replacement plan
+   * @param originalPlan original plan
+   * @param indexPlan replaced plan with index
+   * @return complementIndexPlan integrated plan of indexPlan and complementPlan
+   */
+  private def getComplementIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      originalPlan: LogicalPlan,
+      indexPlan: LogicalPlan): LogicalPlan = {
+    // 1) Replace the location of LogicalRelation with appended files
+    val complementIndexPlan = originalPlan transformDown {
+      case baseRelation @ LogicalRelation(
+            fsRelation @ HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+            baseOutput,
+            _,
+            _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion
+        val updatedOutput =
+          baseOutput.filter(attr => index.schema.fieldNames.contains(attr.name))
+
+        val filesAppended = (location.allFiles
+          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+          .toSet -- index.allSourceFileInfos).toSeq.map(f => new Path(f.name))
+
+        if (filesAppended.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, filesAppended, Map(), None)
+          val newRelation =
+            fsRelation.copy(
+              location = newLocation,
+              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
+              spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
+    }
+
+    if (!originalPlan.equals(complementIndexPlan)) {
```

Comment:
So, it looks like if `filesAppended` computed in `getHybridScanIndexPlan ` is empty, we don't even need to call `getComplementIndexPlan`, no?

Reply:
Yes but it's not always computed in getHybridScanIndexPlan. 
Actually I'd like to compute the current file list & appended files & deleted files for each index once and keep across Rules ..

Reply:
Well, I see that `filesAppended` is computed `if (!useBucketUnion)` in `getHybridScanIndexPlan`, but this function is getting called `if (useBucketUnion)`, so it's always computed anyway?

I was thinking something like:
```scala
var appendedFiles = Nil // This can be updated to a tuple when deletedFiles are introduced.
val replacedPlan = plan transformDown {
  // Always update appendedFiles.
}

if (useBucketUnion && appendFiles.nonEmpty) {
  bucketUnion(spark, index, plan, replacedPlan)
} else {
  replacedPlan
}
```
I think this will make it much easier to understand the flow. (Sorry if I misunderstood the code.)

Reply:
In this way, if we don't pass the appended file list, it's computed twice in both functions.
In order to pass the appended file list to the getComplementIndexPlan, we might need a map <LogicalRelation, appendedfiles> (for multiple relations later).. is it okay to introduce these kind of map here?

Reply:
Oops, yes, I meant to pass the `appendFiles` to the function.

Another reason to cache if we plan to support multiple relations. :) But, no need to introduce the mapping in this PR. Btw, `transformPlanWithIndex` supports only linear LogicalRelation at the moment right (meaning it will break)? I think I left a comment about this pre-requisite, but forgot about it.

Reply:
I updated pre-requisite comment and added `filesAppended`. It looks better to understand the flow 👍

Reply:
While adding 'var files appended', I refactored getComplemetIndex and also applied the Union change as I thought the change is simple enough. 
Sorry for the big diff, but I think this looks better than before.